### PR TITLE
Get NTDS.DIT location on a DC

### PR DESCRIPTION
**Pull request by stufus**
# Overview

This is a really simple module that I put together whilst on site during a penetration test. I needed to quickly find out the location of NTDS.DIT on a domain controller and had to google the registry key for this.

It turns out that a number of NTDS.DIT-grabber modules will read this key, but there was no straightforward way of just getting the confirmed location of the file without attempting to actually download it. I thought this would be useful in itself.

I therefore made this module which, in essence, reads a registry key, looks for the file referenced by it and (if it exists), displays information about it. 

For reference, the registry key is HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\NTDS\Parameters and there are a number of values that can be chosen. The most useful one is "DSA Database file" which is the absolute path to NTDS.DIT.
# Testing

```
msf post(ntds_location) > run

NTDS.DIT is located at: F:\AD\Client\NTDS\ntds.dit
      Size: 16793600 bytes
   Created: 2012-09-03 11:34:18 +0100
  Modified: 2015-09-24 13:03:08 +0100
  Accessed: 2012-09-03 11:34:18 +0100
[*] Post module execution completed
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
What about ```post/windows/gather/credentials/domain_hashdump```?

**Comment by wvu-r7** [comment]
Hmm, how about a module that optionally downloads it?

File: modules/post/windows/gather/ntds_location.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                      'Author'          => ['Stuart Morgan <stuart.morgan[at]mwrinfosecurity.com>']
+                     ))
+  end
+
+  def run
+    # Find the NTDS.DIT location in the registry
+    ntds = registry_getvaldata("HKLM\\SYSTEM\\CurrentControlSet\\Services\\NTDS\\Parameters", "DSA Database file").to_s
+    unless ntds
+      print_error("Unable to find the NTDS.DIT location.")
+      return
```
Comment (OJ) [inline modules/post/windows/gather/ntds_location.rb]: `fail_with` ?

File: modules/post/windows/gather/ntds_location.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                      'Name'            => "NTDS.DIT Location Module",
+                      'Description'     => %q{
+                        This module will find the location of the NTDS.DIT file (from the registry), check that it exists
+                        and display it on the screen. Useful if you wish to manually acquire the file using ntdsutil or vss.
+                       },
+                      'License'         => MSF_LICENSE,
+                      'Platform'        => ['win'],
+                      'SessionTypes'    => ['meterpreter'],
+                      'Author'          => ['Stuart Morgan <stuart.morgan[at]mwrinfosecurity.com>']
+                     ))
```
Comment (OJ) [inline modules/post/windows/gather/ntds_location.rb]: indenting is a little funky here.

**Comment by OJ** [comment]
@stufus: So in what cases would you want to locate the file without downloading it? Cheers!

@wvu-r7: Aren't there modules that do this already?

**Comment by wvu-r7** [comment]
@OJ Not the downloading part, just the using part? Feel like this module should do a bit more than ```stat``` the file.

**Comment by stufus** [comment]
@OJ: A very fair question. In this specific case, I wanted to get the file size because the link between me and the compromised box was very slow. 

@wvu-r7: There are two already that do this, windows/gather/credentials/domain_hashdump.rb and auxiliary/admin/smb/psexec_ntdsgrab.rb (which I'm guessing you already know) :) Three if you include post/windows/gather/file_from_raw_ntfs.rb.

psexec_ntdsgrab.rb works using volume shadow services to snapshot the drive and pull down the file and retrieves it over SMB.

domain_hashdump.rb will either use ntdsutil or vss (depending on version of Windows) to get the same files. However, (and I may have misunderstood the code here), I _think_ that this actually does the parsing on the DC itself, which could put significant load on the DC. In this case, the .dit file was around 800Mb, so I downloaded it manually. What threw me when I was on the test was that it was in a non-standard location and I couldn't lay my hands on a "where is it and how big is it" module.

Initially I was going to write a module that would obtain and download ntds.git and the SYSTEM registry hive, but parse them locally and not remotely. However, that would be the 3rd module that did that and the other two are perfectly good most of the time.

Appreciate the questions and I won't lose any sleep if you close this off without merging - it was just handy for me in a real-world situation and I thought I'd share :)

**Comment by wvu-r7** [comment]
@stufus: I shouldn't have deleted my comment about ```post/windows/gather/credentials/domain_hashdump```, lol.

**Comment by stufus** [comment]
@wvu-r7: what comment? :) haha

It was one of those things mate - it was a quick and dirty fix that took a few minutes to put together, thought I'd offer it up just in case it could be useful. I went hunting to see if I could just add a 'check' or 'show-location' command to any of the others because the module is a bit heavyweight for something so simple but couldn't see how to do it...

**Comment by OJ** [comment]
@stufus: I didn't intend for my question to sound dismissive ;) I'm genuinely interested in the use case. I'm just putting myself in the shoes of the user and asking "after I stat the file, what would I do next?"

**Comment by stufus** [comment]
@OJ: you weren't dismissive at all mate, I didn't want to take up your time with semi-worthless modules, that's all :)

**Comment by OJ** [comment]
@stufus: "worthless" is not a word I would associate with your efforts.

**Comment by stufus** [comment]
@OJ: haha thanks mate! :) All good fun

**Comment by wvu-r7** [comment]
Well, if we're not gonna download it, how about we delete it? >:P

**Comment by stufus** [comment]
@OJ: So as an answer, the user would then ask themselves which method they want to use to download said file (taking into account the attending circumstances) and then either manually do it or use a module, or perhaps use it as a way of verifying whether the server is in fact a DC....or something....

@wvu-r7 Why not :) Or overwrite it with a huge ascii art metasploit motd and see if anyone experiences domain auth problems....or overwrite it with those wavs that appeared in my more recent git pull....

**Comment by wvu-r7** [comment]
And we have modules that do both those things, just not ONLY those things.

**Comment by OJ** [comment]
`use post/windows/destroy/domain_controller`

**Comment by wvu-r7** [comment]
@OJ Your profile pic...

**Comment by OJ** [comment]
Yeah, with that kind of comment I guess it looks rather sinister! :see_no_evil:

**Comment by stufus** [comment]
well I suppose you could always make it use post/windows/retire/domain_controller or something...

**Comment by wvu-r7** [comment]
"Retire" sounds much more forgiving.

**Comment by wvu-r7** [comment]
So, what do we want to do here? We happy with what it does now?

File: modules/post/windows/gather/ntds_location.rb (hunk 1 above)
Comment (wvu-r7) [inline modules/post/windows/gather/ntds_location.rb]: Since this isn't really a module failure, I think it's fine.

**Comment by wvu-r7** [comment]
@stufus && @OJ: Changed here: https://github.com/rapid7/metasploit-framework/commit/06a2bb53bd145248ca25c6dfdc124cf8ee8456e3.

**Pull request merged**